src/mds/Server.cc

```diff
@@ -1274,7 +1275,7 @@ void Server::handle_client_reconnect(const cref_t<MClientReconnect> &m)
 	  << (m->has_more() ? " (more)" : "") << dendl;
   client_t from = m->get_source().num();
   Session *session = mds->get_session(m);
-  if (!session) {
+  if (!session || !session->is_open()) {
```

Comment:
I don't think this one is correct. It causes mount failures:

````
2019-10-03T18:20:36.349 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi060.stderr:2019-10-03T18:20:36.347+0000 7fd017016f40 -1 init, newargv = 0x5646365635d0 newargc=7
2019-10-03T18:20:36.349 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi060.stderr:ceph-fuse[14299]: starting ceph client
2019-10-03T18:20:36.354 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi060.stderr:2019-10-03T18:20:36.351+0000 7fcffb7fe700 -1 client.4599 mds.0 rejected us (sessionless)
2019-10-03T18:20:36.354 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi060.stderr:ceph-fuse[14299]: ceph mount failed with (1) Operation not permitted
2019-10-03T18:20:36.524 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi060.stderr:daemon-helper: command failed with exit status 1
````

http://pulpito.ceph.com/pdonnell-2019-10-03_18:03:01-fs-wip-pdonnell-testing-20191003.135856-distro-basic-smithi/4354213/

Reply:
> I don't think this one is correct. It causes mount failures:
> 
> ```
> 2019-10-03T18:20:36.349 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi060.stderr:2019-10-03T18:20:36.347+0000 7fd017016f40 -1 init, newargv = 0x5646365635d0 newargc=7
> 2019-10-03T18:20:36.349 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi060.stderr:ceph-fuse[14299]: starting ceph client
> 2019-10-03T18:20:36.354 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi060.stderr:2019-10-03T18:20:36.351+0000 7fcffb7fe700 -1 client.4599 mds.0 rejected us (sessionless)
> 2019-10-03T18:20:36.354 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi060.stderr:ceph-fuse[14299]: ceph mount failed with (1) Operation not permitted
> 2019-10-03T18:20:36.524 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi060.stderr:daemon-helper: command failed with exit status 1
> ```
> 
> http://pulpito.ceph.com/pdonnell-2019-10-03_18:03:01-fs-wip-pdonnell-testing-20191003.135856-distro-basic-smithi/4354213/

Changed CEPH_SESSION_REJECT to CEPH_SESSION_CLOSE firstly, the reason refers to https://github.com/ceph/ceph/pull/27256. Null session shouldn't be rejected directly just because of time out in reconnect time. 
This also resolves fuse mount failure.
But I am not clear why fuse-mount uses reconnect msg .

Reply:
I think it's because:
1. client sends open request to mds
2. mds restarts
3. client don't know if mds has opened its session, so it sends reconnect.

So, we should send session close back for the  !session->open() case

Reply:
> I think it's because:
> 
> 1. client sends open request to mds
> 2. mds restarts
> 3. client don't know if mds has opened its session, so it sends reconnect.
> 
> So, we should send session close back for the !session->open() case

OK ,changed